crates/gitbutler-project-store/src/migrations/migrator.rs

```diff
@@ -0,0 +1,204 @@
+            up: |connection| {
+                connection.execute(
+                    "
+                CREATE TABLE testaroni (
+                    id BLOB PRIMARY KEY, -- A V4 UUID
+                    name TEXT NOT NULL
+                )
+                ",
+                    [],
+                )?;
+
+                Ok(())
+            },
+        }
+    }
+
+    /// Depends on `first_migration`
+    fn second_migration() -> Migration {
+        Migration {
+            name: "second".to_string(),
+            up: |connection| {
+                connection.execute("ALTER TABLE testaroni ADD potatoes TEXT", [])?;
+
+                Ok(())
+            },
+        }
+    }
+
+    #[test]
+    /// Testing the `find_applied_migrations` function.
+    /// Ensuring that it creates the `migrations` table if it doesn't exist
+    fn find_applied_migrations_creates_migrations_table_if_not_exists() {
+        let mut connection = Connection::open_in_memory().unwrap();
+
+        let migrator = Migrator::new(&mut connection);
+
+        // Try to find unapplied migrations
+        migrator.find_applied_migrations().unwrap();
+
+        let statement = connection
+            .prepare("SELECT name FROM sqlite_master WHERE type='table' AND name='migrations'")
+            .unwrap();
+        assert_eq!(statement.column_count(), 1);
+    }
+
+    #[test]
+    /// Testing the `find_applied_migrations` function.
+    /// Assuming the `migrations` table exists (which we create by calling
+    /// `find_applied_migrations), we expect that any values inserted into
+    /// the table will get returned when we call it again.
+    fn find_applied_migrations_lists_all_entries_in_table() {
+        let mut connection = Connection::open_in_memory().unwrap();
+
+        // Call find_applied_migrations in order to create the migrations table
+        {
+            let migrator = Migrator::new(&mut connection);
+            migrator.find_applied_migrations().unwrap();
+        }
+
+        // Insert some entries into the migrations table
+        connection
+            .execute(
+                "INSERT INTO migrations (id, name) VALUES (?1, ?2)",
+                (Uuid::new_v4(), "base".to_string()),
+            )
+            .unwrap();
+
+        connection
+            .execute(
+                "INSERT INTO migrations (id, name) VALUES (?1, ?2)",
+                (Uuid::new_v4(), "other".to_string()),
+            )
+            .unwrap();
+
+        {
+            let migrator = Migrator::new(&mut connection);
+            let results = migrator.find_applied_migrations().unwrap();
+            assert_eq!(results.len(), 2);
+            assert!(results.contains(&"base".to_string()));
+            assert!(results.contains(&"other".to_string()));
+        }
+    }
+
+    #[test]
+    /// Testing the `migrate` function.
+    /// When given a list of migrations to run, it will perform the migrations in order
+    fn migrate_applies_migrations_in_order() {
```

Comment:
I recommend to not test `find_applied_migrations()` as it needs to know a lot of details. 
Instead, let `migrate()` return enough information to know which migrations were applied. That way there can be a 'journey' that does what happens typically, empty database, then the first migration, then the second, with idempotency checks sprinkled in.

Tests that are considered good because they don't panic are an indicator that some more useful information could be returned organically  to make what happens more observable.